book-content/chapters/08-classes.md

```diff
@@ -0,0 +1,1044 @@
+
+```typescript
+loopFindingJazzRecords.printAlbumInfo();
+
+// Output: Loop Finding Jazz Records by Jan Jelinek, released in 2001.
+```
+
+#### Arrow Functions or Class Methods?
+
+Arrow functions and class methods do differ in their behavior. The difference is the way that `this` is handled.
+
+This is runtime JavaScript behavior, so slightly outside the scope of this book. But in the interest of helpfulness, here's an example:
+
+```typescript
+class MyClass {
+  location = "Class";
+
+  arrow = () => {
+    console.log("arrow", this);
+  };
+
+  method() {
+    console.log("method", this);
+  }
+}
+
+const myObj = {
+  location: "Object",
+  arrow: new MyClass().arrow,
+  method: new MyClass().method,
+};
+
+myObj.arrow(); // { location: 'Class' }
+myObj.method(); // { location: 'Object' }
+```
+
+In the `arrow` method, `this` is bound to the instance of the class where it was defined. In the `method` method, `this` is bound to the object where it was called.
+
+This can be a bit of a gotcha when working with classes, whether in JavaScript or TypeScript.
+
+## Class Inheritance
+
+Similar to how we can extend types and interfaces, we can also extend classes in TypeScript. This allows you to create a hierarchy of classes that can inherit properties and methods from one another, making your code more organized and reusable.
+
+For this example, we'll go back to our basic `Album` class that will act as our base class:
+
+```typescript
+class Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+
+  constructor(opts: { title: string; artist: string; releaseYear: number }) {
+    this.title = title;
+    this.artist = artist;
+    this.releaseYear = releaseYear;
+  }
+
+  displayInfo() {
+    console.log(
+      `${this.title} by ${this.artist}, released in ${this.releaseYear}.`,
+    );
+  }
+}
+```
+
+The goal is to create a `SpecialEditionAlbum` class that extends the `Album` class and adds a `bonusTracks` property.
+
+### Extending a Class
+
+The first step is to use the `extends` keyword to create the `SpecialEditionAlbum` class:
+
+```typescript
+class SpecialEditionAlbum extends Album {}
+```
+
+Once the `extends` keyword is added, any new properties or methods added to the `SpecialEditionAlbum` class will be in addition to what it inherits from the `Album` class. For example, we can add a `bonusTracks` property to the `SpecialEditionAlbum` class:
+
+```typescript
+class SpecialEditionAlbum extends Album {
+  bonusTracks: string[];
+}
+```
+
+Next, we need to add a constructor that includes all of the properties from the `Album` class as well as the `bonusTracks` property. There are a couple of important things to note about the constructor when extending a class.
+
+First, the arguments to the constructor should match the shape used in the parent class. In this case, that's an `opts` object with the properties of the `Album` class along with the new `bonusTracks` property.
```

Comment:
There is no such requirement, I wouldn't recommend deviating too far but this is perfectly valid code for TS:
```ts
class Base1 {
  constructor(arg: number) {}
}
class Cls1 extends Base1 {
  constructor(arg: string) {
    super(parseInt(arg));
  }
}

class Base2 {
  constructor(arg: { foo: string }) {}
}
class Cls2 extends Base2 {
  constructor(arg: { bar: string }) {
    super({ foo: arg.bar });
  }
}
```